apiserver/facades/client/storage/register.go

```diff
@@ -16,32 +17,36 @@ import (
 // Register is called to expose a package of facades onto a given registry.
 func Register(registry facade.FacadeRegistry) {
 	registry.MustRegister("Storage", 6, func(stdCtx context.Context, ctx facade.ModelContext) (facade.Facade, error) {
-		return newStorageAPI(ctx) // modify Remove to support force and maxWait; add DetachStorage to support force and maxWait.
+		return newStorageAPI(stdCtx, ctx) // modify Remove to support force and maxWait; add DetachStorage to support force and maxWait.
 	}, reflect.TypeOf((*StorageAPI)(nil)))
 }
 
 // newStorageAPI returns a new storage API facade.
-func newStorageAPI(ctx facade.ModelContext) (*StorageAPI, error) {
+func newStorageAPI(stdCtx context.Context, ctx facade.ModelContext) (*StorageAPI, error) {
 	st := ctx.State()
-	model, err := st.Model()
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
 
 	domainServices := ctx.DomainServices()
 	storageAccessor, err := getStorageAccessor(st)
 	if err != nil {
 		return nil, errors.Annotate(err, "getting backend")
 	}
 
+	modelType, err := domainServices.Model().ModelType(stdCtx, ctx.ModelUUID())
+	if err != nil {
+		return nil, errors.Annotate(err, "getting model backend")
+	}
+
 	authorizer := ctx.Auth()
 	if !authorizer.AuthClient() {
 		return nil, apiservererrors.ErrPerm
 	}
 
 	storageService := domainServices.Storage()
 	return NewStorageAPI(
-		stateShim{st}, model.Type(),
+		names.NewControllerTag(ctx.ControllerUUID()),
```

Comment:
Same comments as the other PR's these should be the uuid's and not tags.

Reply:
Facades deal with tags - tags are used for permission checking inside the facade, plus the facade pulls tags off the wire etc. The tags never leave the facade. I wish the permission checks did not use tags, but while they do, doing the allocation once up front is good.

Reply:
changed to uuids